src/pybind/mgr/ssh/module.py

```diff
@@ -0,0 +1,725 @@
+        keyring = self._get_bootstrap_key("mgr")
+        keyring_path = "/var/lib/ceph/bootstrap-mgr/ceph.keyring"
+        conn.remote_module.write_keyring(keyring_path, keyring)
+        return keyring_path
+
+    def _bootstrap_osd(self, conn):
+        """
+        Bootstrap an osd.
+
+          1. install a copy of ceph.conf
+          2. install the osd bootstrap key
+
+        :param conn: remote host connection
+        """
+        self._ensure_ceph_conf(conn)
+        keyring = self._get_bootstrap_key("osd")
+        keyring_path = "/var/lib/ceph/bootstrap-osd/ceph.keyring"
+        conn.remote_module.write_keyring(keyring_path, keyring)
+        return keyring_path
+
+    def _hostname_to_store_key(self, host):
+        return "{}.{}".format(self._STORE_HOST_PREFIX, host)
+
+    def _get_hosts(self, wanted=None):
+        if wanted:
+            hosts_info = []
+            for host in wanted:
+                key = self._hostname_to_store_key(host)
+                info = self.get_store(key)
+                if info:
+                    hosts_info.append((key, info))
+        else:
+            hosts_info = six.iteritems(self.get_store_prefix(self._STORE_HOST_PREFIX))
+
+        return list(map(lambda kv: (kv[0], json.loads(kv[1])), hosts_info))
+
+    def add_host(self, host):
+        """
+        Add a host to be managed by the orchestrator.
+
+        :param host: host name
+        :param labels: host labels
+        """
+        def run(host):
+            key = self._hostname_to_store_key(host)
+            self.set_store(key, json.dumps({
+                "host": host,
+                "inventory": None,
+                "last_inventory_refresh": None
+            }))
+            return "Added host '{}'".format(host)
+
+        return SSHWriteCompletion(
+            self._worker_pool.apply_async(run, (host,)))
+
+    def remove_host(self, host):
+        """
+        Remove a host from orchestrator management.
+
+        :param host: host name
+        """
+        def run(host):
+            key = self._hostname_to_store_key(host)
+            self.set_store(key, None)
+            return "Removed host '{}'".format(host)
+
+        return SSHWriteCompletion(
+            self._worker_pool.apply_async(run, (host,)))
+
+    def get_hosts(self):
+        """
+        Return a list of hosts managed by the orchestrator.
+
+        Notes:
+          - skip async: manager reads from cache.
+
+        TODO:
+          - InventoryNode probably needs to be able to report labels
+        """
+        nodes = []
+        for key, host_info in self._get_hosts():
+            node = orchestrator.InventoryNode(host_info["host"], [])
+            nodes.append(node)
+        return SSHReadCompletionReady(nodes)
+
+    def _get_device_inventory(self, host):
+        """
+        Query storage devices on a remote node.
+
+        :return: list of InventoryDevice
+        """
+        conn = self._get_connection(host)
```

Comment:
AFAIC, you can use _get_connection like so: 
```python
with self._get_connection(host) as conn:
    ceph_volume_executable = self._executable_path(conn, 'ceph-volume')
    ....
```
Then, there is no need to manually call `conn.exit()`

Reply:
nice. i didn't notice that connection implemented `__exit__`